// react-native-worklets/lib/module/WorkletsModule/NativeWorklets.js
/* eslint-disable @typescript-eslint/unbound-method */
'use strict';

import { WorkletsTurboModule } from "../specs/index.js";
import { WorkletsError } from "../WorkletsError.js";
export function createNativeWorkletsModule() {
  return new NativeWorklets();
}
class NativeWorklets {
  #workletsModuleProxy;
  #shareableUndefined;
  #shareableNull;
  #shareableTrue;
  #shareableFalse;
  constructor() {
    if (global.__workletsModuleProxy === undefined) {
      WorkletsTurboModule?.installTurboModule();
    }
    if (global.__workletsModuleProxy === undefined) {
      throw new WorkletsError(`Native part of Worklets doesn't seem to be initialized.
See https://docs.swmansion.com/react-native-reanimated/docs/guides/troubleshooting#native-part-of-reanimated-doesnt-seem-to-be-initialized for more details.`);
    }
    this.#workletsModuleProxy = {
      scheduleOnUI: global.__workletsModuleProxy.scheduleOnUI,
      scheduleOnRuntime: global.__workletsModuleProxy.scheduleOnRuntime,
      executeOnUIRuntimeSync: global.__workletsModuleProxy.executeOnUIRuntimeSync,
      createWorkletRuntime: global.__workletsModuleProxy.createWorkletRuntime,
      makeShareableClone: global.__workletsModuleProxy.makeShareableClone,
      makeShareableString: global.__workletsModuleProxy.makeShareableString,
      makeShareableNumber: global.__workletsModuleProxy.makeShareableNumber,
      makeShareableBoolean: global.__workletsModuleProxy.makeShareableBoolean,
      makeShareableBigInt: global.__workletsModuleProxy.makeShareableBigInt,
      makeShareableUndefined: global.__workletsModuleProxy.makeShareableUndefined,
      makeShareableNull: global.__workletsModuleProxy.makeShareableNull
    };
    this.#shareableNull = this.#workletsModuleProxy.makeShareableNull();
    this.#shareableUndefined = this.#workletsModuleProxy.makeShareableUndefined();
    this.#shareableTrue = this.#workletsModuleProxy.makeShareableBoolean(true);
    this.#shareableFalse = this.#workletsModuleProxy.makeShareableBoolean(false);
  }
  makeShareableClone(value, shouldPersistRemote, nativeStateSource) {
    return this.#workletsModuleProxy.makeShareableClone(value, shouldPersistRemote, nativeStateSource);
  }
  makeShareableString(str) {
    return this.#workletsModuleProxy.makeShareableString(str);
  }
  makeShareableNumber(num) {
    return this.#workletsModuleProxy.makeShareableNumber(num);
  }
  makeShareableBoolean(bool) {
    return bool ? this.#shareableTrue : this.#shareableFalse;
  }
  makeShareableBigInt(bigInt) {
    return this.#workletsModuleProxy.makeShareableBigInt(bigInt);
  }
  makeShareableUndefined() {
    return this.#shareableUndefined;
  }
  makeShareableNull() {
    return this.#shareableNull;
  }
  scheduleOnUI(shareable) {
    return this.#workletsModuleProxy.scheduleOnUI(shareable);
  }
  executeOnUIRuntimeSync(shareable) {
    return this.#workletsModuleProxy.executeOnUIRuntimeSync(shareable);
  }
  createWorkletRuntime(name, initializer) {
    return this.#workletsModuleProxy.createWorkletRuntime(name, initializer);
  }
  scheduleOnRuntime(workletRuntime, shareableWorklet) {
    return this.#workletsModuleProxy.scheduleOnRuntime(workletRuntime, shareableWorklet);
  }
}
